src/test/java/com/arpnetworking/utility/RegexAndMapReplacerTest.java

```diff
@@ -0,0 +1,199 @@
+        final String replace = "this is a ${g1} pattern called ${g2}";
+        final String expected = "this is a test pattern called foo";
+        testExpression(pattern, input, replace, expected, ImmutableMap.of("g1", "test", "g2", "foo"));
+    }
+
+    @Test
+    public void testSingleMatchPartialMultipleVariableWithEscapeReplace() {
+        final Pattern pattern = Pattern.compile("test/pattern/foo");
+        final String input = "test/pattern/foo";
+        final String replace = "this is a ${g1} pattern \\\\called\\\\ ${g2}";
+        final String expected = "this is a test pattern \\called\\ foo";
+        testExpression(pattern, input, replace, expected, ImmutableMap.of("g1", "test", "g2", "foo"));
+    }
+
+    @Test
+    public void testSingleMatchPartialMultipleVariableWithEscapeTokenReplace() {
+        final Pattern pattern = Pattern.compile("test/pattern/foo");
+        final String input = "test/pattern/foo";
+        final String replace = "this is a ${\\\\g1} pattern called ${g2}";
+        final String expected = "this is a test pattern called foo";
+        testExpression(pattern, input, replace, expected, ImmutableMap.of("\\g1", "test", "g2", "foo"));
+    }
+
+    @Test
+    public void testSingleMatchPartialMultipleGroupNameOverridesVariablesReplace() {
+        final Pattern pattern = Pattern.compile("(?<g1>test)/pattern/(?<g2>foo)");
+        final String input = "test/pattern/foo";
+        final String replace = "this is a ${g1} pattern called ${g2}";
+        final String expected = "this is a test pattern called foo";
+        testExpression(pattern, input, replace, expected, ImmutableMap.of("g1", "bad", "g2", "value"));
+    }
+
+    @Test
+    public void testMultipleMatchFullStaticReplace() {
+        final Pattern pattern = Pattern.compile("test");
+        final String input = "testtest";
+        final String replace = "replace";
+        final String expected = "replacereplace";
+        testExpression(pattern, input, replace, expected, Collections.emptyMap());
+    }
+
+    @Test
+    public void testMultipleMatchPartialStaticReplace() {
+        final Pattern pattern = Pattern.compile("test");
+        final String input = "test string test";
+        final String replace = "replace";
+        final String expected = "replace string replace";
+        testExpression(pattern, input, replace, expected, Collections.emptyMap());
+    }
+
+    private void testExpression(final Pattern pattern, final String input, final String replace, final String expected,
+            final Map<String, String> variables) {
+        final String result = RegexAndMapReplacer.replaceAll(pattern.matcher(input), input, replace, variables);
+        Assert.assertEquals(expected, result);
+        try {
+            final String stockResult = pattern.matcher(input).replaceAll(replace);
+            Assert.assertEquals(expected, stockResult);
+        } catch (final IllegalArgumentException ignored) { }
+    }
+}
```

Comment:
We should take this opportunity to document the mapping source in the README.md.